### PR TITLE
Detect multiple concurent writes to Dict + remove dead code

### PR DESCRIPTION
**Pull request by petvana**
Thanks to #38180, the removed code seems to be dead because no finalizers should be used anymore to modify dictionaries (it was dangerous). Furthermore, it may help users to detect illegal concurrent writes, since it doesn't recurse and have new error message. There should be no, or even a positive, performance effect.

Btw, this seems to be the last missing step to allow manual shrinkage at `sizehint!`
https://github.com/JuliaLang/julia/blob/89a613b5978ecf06566f5778253ca18c3e3fc2e7/base/dict.jl#L251-L256

**Master**
``` julia
$ julia -t 2

julia> d = Dict()
Dict{Any, Any}()

julia> Threads.@threads for i = 1:1000; d[i] = i; end
#... Hangs for ... so long ... and I terminated the loop
```

**PR**
``` julia
$ julia -t 2

julia> d = Dict()
Dict{Any, Any}()

julia> Threads.@threads for i = 1:1000; d[i] = i; end
ERROR: TaskFailedException
Stacktrace:
 [1] wait
   @ ./task.jl:345 [inlined]
 [2] threading_run(fun::var"#23#threadsfor_fun#5"{var"#23#threadsfor_fun#4#6"{UnitRange{Int64}}}, static::Bool)
   @ Base.Threads ./threadingconstructs.jl:38
 [3] top-level scope
   @ ./threadingconstructs.jl:89

    nested task error: AssertionError: Muliple concurent writes to Dict detected!
    Stacktrace:
     [1] rehash!(h::Dict{Any, Any}, newsz::Int64)
       @ Base ./dict.jl:235
     [2] _setindex!
       @ ./dict.jl:382 [inlined]
     [3] setindex!(h::Dict{Any, Any}, v::Any, key::Int64)
       @ Base ./dict.jl:419
     [4] macro expansion
       @ ./REPL[2]:1 [inlined]
     [5] (::var"#23#threadsfor_fun#5"{var"#23#threadsfor_fun#4#6"{UnitRange{Int64}}})(tid::Int64; onethread::Bool)
       @ Main ./threadingconstructs.jl:84
     [6] #23#threadsfor_fun
       @ ./threadingconstructs.jl:52 [inlined]
     [7] (::Base.Threads.var"#1#2"{var"#23#threadsfor_fun#5"{var"#23#threadsfor_fun#4#6"{UnitRange{Int64}}}, Int64})()
       @ Base.Threads ./threadingconstructs.jl:30
```

### CODE REVIEW

File: base/dict.jl [hunk 1]
```diff
@@ -237,7 +232,7 @@ end
     h.count = count
     h.ndel = 0
     h.maxprobe = maxprobe
-    @assert h.age == age0
+    @assert h.age == age0 "Muliple concurent writes to Dict detected!"
```
Comment (vtjnash) [inline base/dict.jl:235]: move this above the assignments, so we only do the mutation if things are not broken?

**Comment by vtjnash** [comment]
This seems good. Does it improve TTFP too? This call used to cause a lot of inference latency.

**Comment by vtjnash** [comment]
(I am fairly sure that comment is also very old, and refers to a time when we assumed that doubling the size meant eliminating hash collisions–a false assumption we deleted long ago)

File: base/dict.jl (hunk 1 above)
Comment (petvana) [reply base/dict.jl]: It's only a sanity check for now and thus can be changed anywhere. I've also added one aging for more safety.

**Comment by petvana** [comment]
> This seems good. Does it improve TTFP too? This call used to cause a lot of inference latency.

Unfortunately no, TTFP seems unchanged.

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44778#issuecomment-1081303121) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/3d00d0a_vs_c919811/report.md).

File: test/dict.jl [hunk 2]
```diff
@@ -1247,3 +1247,16 @@ end
...
+@testset "Detect multiple concurent writes" begin
+    if Threads.nthreads() > 1
+        @test_throws TaskFailedException let d = Dict()
+            Threads.@threads for i = 1:1000
+                d[i] = i
+            end
+        end
+    else
+        # Multiple threads not enabled.
+        @test false
```
Comment (petvana) [inline test/dict.jl:1260]: Not sure, if multiple threads are enabled in CI. Just to make sure.

Comment (petvana) [reply test/dict.jl:1260]: Any idea on how to enable CI with `Threads.nthreads() != 1`? Going through other tests without and not finding the right way.

Comment (petvana) [reply test/dict.jl:1260]: Does it mean that the following test is disabled now in CI? https://github.com/JuliaLang/julia/blob/897c3cfdda888c3149724531e07cdd4f0fede895/test/threads_exec.jl#L30-L32

**Comment by petvana** [comment]
> The test seems fundamentally flawed since there is no guarantee you actually will run into a data race, so you might get nondeterministic test failures.

Fair point, thanks. Although the probability would be very low for the test to fail for larger dictionaries (based on how long rehashing takes), it may still happen. So, I removed the test, and PR is ready to be merged.

**Review by simeonschaub — APPROVED** [review]

File: base/dict.jl [hunk 3]
```diff
@@ -223,22 +223,19 @@ end
...
-
-            if h.age != age0
-                # if `h` is changed by a finalizer, retry
-                return rehash!(h, newsz)
-            end
         end
     end
 
+    @assert h.age == age0 "Muliple concurent writes to Dict detected!"
+
```
Comment (giordano) [inline base/dict.jl:230]:
```suggestion

```

File: base/dict.jl [hunk 4]
```diff
@@ -223,22 +223,19 @@ end
...
+    h.age += 1
     h.slots = slots
     h.keys = keys
     h.vals = vals
     h.count = count
     h.ndel = 0
     h.maxprobe = maxprobe
-    @assert h.age == age0
-
+
```
Comment (giordano) [inline base/dict.jl:238]:
```suggestion

```

**Review by giordano — CHANGES_REQUESTED** [review]

**Review by giordano — APPROVED** [review]

**Pull request merged**